In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
df=pd.read_csv('/content/Billing_Data.csv')
df1=pd.read_csv('/content/Patient_Data.csv')

In [ ]:
df

,PatientID,InsuranceCovered,FinalAmount
0,101,2000,3000
1,102,1500,3500
2,103,2500,5000
3,104,3000,3200
4,105,1000,4000


In [ ]:
df1

,PatientID,Name,Department,Doctor,BillAmount,ReceptionistID,CheckInTime
0,101,Alice,Cardiology,Dr. Smith,5000.0,1,2023-01-10 09:00
1,102,Bob,Neurology,Dr. John,NaN,2,2023-01-11 10:30
2,103,Charlie,Orthopedics,Dr. Lee,7500.0,1,2023-01-12 11:00
3,104,David,Cardiology,Dr. Smith,6200.0,3,2023-01-13 12:00
4,105,Eva,Dermatology,Dr. Rose,NaN,2,2023-01-14 08:45
5,101,Alice,Cardiology,Dr. Smith,5000.0,1,2023-01-10 09:00


In [ ]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   PatientID       6 non-null      int64  
 1   Name            6 non-null      object 
 2   Department      6 non-null      object 
 3   Doctor          6 non-null      object 
 4   BillAmount      4 non-null      float64
 5   ReceptionistID  6 non-null      int64  
 6   CheckInTime     6 non-null      object 
dtypes: float64(1), int64(2), object(4)
memory usage: 468.0+ bytes


In [ ]:
df1[['PatientID', 'Department', 'Doctor', 'BillAmount']]

,PatientID,Department,Doctor,BillAmount
0,101,Cardiology,Dr. Smith,5000.0
1,102,Neurology,Dr. John,NaN
2,103,Orthopedics,Dr. Lee,7500.0
3,104,Cardiology,Dr. Smith,6200.0
4,105,Dermatology,Dr. Rose,NaN
5,101,Cardiology,Dr. Smith,5000.0


In [ ]:
# Drop columns
df1 = df1.drop(columns=["ReceptionistID", "CheckInTime"])

print(df1)

   PatientID     Name   Department     Doctor  BillAmount
0        101    Alice   Cardiology  Dr. Smith      5000.0
1        102      Bob    Neurology   Dr. John         NaN
2        103  Charlie  Orthopedics    Dr. Lee      7500.0
3        104    David   Cardiology  Dr. Smith      6200.0
4        105      Eva  Dermatology   Dr. Rose         NaN
5        101    Alice   Cardiology  Dr. Smith      5000.0


In [ ]:
# Total bill amount for each department
total_bill = df1.groupby("Department")["BillAmount"].sum()

print(total_bill)

Department
Cardiology     16200.0
Dermatology        0.0
Neurology          0.0
Orthopedics     7500.0
Name: BillAmount, dtype: float64


In [ ]:
df1.duplicated().sum()

np.int64(1)

In [ ]:
# Remove duplicate PatientID records
df = df.drop_duplicates(subset="PatientID")

print(df)

   PatientID  InsuranceCovered  FinalAmount
0        101              2000         3000
1        102              1500         3500
2        103              2500         5000
3        104              3000         3200
4        105              1000         4000


In [ ]:
# Find the mean bill amount
mean_bill = df1["BillAmount"].mean()

# Fill missing values with the mean
df1["BillAmount"] = df1["BillAmount"].fillna(mean_bill)

print(df1)

   PatientID     Name   Department     Doctor  BillAmount
0        101    Alice   Cardiology  Dr. Smith      5000.0
1        102      Bob    Neurology   Dr. John      5925.0
2        103  Charlie  Orthopedics    Dr. Lee      7500.0
3        104    David   Cardiology  Dr. Smith      6200.0
4        105      Eva  Dermatology   Dr. Rose      5925.0
5        101    Alice   Cardiology  Dr. Smith      5000.0


In [ ]:
merged_df = pd.merge(df1, df, on="PatientID")

print(merged_df)

   PatientID     Name   Department     Doctor  BillAmount  InsuranceCovered  \
0        101    Alice   Cardiology  Dr. Smith      5000.0              2000   
1        102      Bob    Neurology   Dr. John      5925.0              1500   
2        103  Charlie  Orthopedics    Dr. Lee      7500.0              2500   
3        104    David   Cardiology  Dr. Smith      6200.0              3000   
4        105      Eva  Dermatology   Dr. Rose      5925.0              1000   
5        101    Alice   Cardiology  Dr. Smith      5000.0              2000   

   FinalAmount  
0         3000  
1         3500  
2         5000  
3         3200  
4         4000  
5         3000  


In [ ]:
new_patients = pd.DataFrame({
    "PatientID": [103, 104],
    "Name": ["Amit", "Sneha"],
    "Department": ["Orthopedic", "Cardiology"]
})

df = pd.concat([df, new_patients], ignore_index=True)

print(df)

   PatientID  InsuranceCovered  FinalAmount   Name  Department
0        101            2000.0       3000.0    NaN         NaN
1        102            1500.0       3500.0    NaN         NaN
2        103            2500.0       5000.0    NaN         NaN
3        104            3000.0       3200.0    NaN         NaN
4        105            1000.0       4000.0    NaN         NaN
5        103               NaN          NaN   Amit  Orthopedic
6        104               NaN          NaN  Sneha  Cardiology


In [ ]:
billing_info = pd.DataFrame({
    "InsuranceCovered": [2000, 3000, 1500],
    "FinalAmount": [3000, 4000, 2500]
})

df = pd.concat([df, billing_info], axis=1)

print(df)

   PatientID  InsuranceCovered  FinalAmount   Name  Department  \
0        101            2000.0       3000.0    NaN         NaN   
1        102            1500.0       3500.0    NaN         NaN   
2        103            2500.0       5000.0    NaN         NaN   
3        104            3000.0       3200.0    NaN         NaN   
4        105            1000.0       4000.0    NaN         NaN   
5        103               NaN          NaN   Amit  Orthopedic   
6        104               NaN          NaN  Sneha  Cardiology   

   InsuranceCovered  FinalAmount  
0            2000.0       3000.0  
1            3000.0       4000.0  
2            1500.0       2500.0  
3               NaN          NaN  
4               NaN          NaN  
5               NaN          NaN  
6               NaN          NaN  


**excepected outcomes**:
1. Final cleaned dataset with accurate billing info
Meaning

After cleaning the data:

Unnecessary columns are removed.
Duplicate patient records are removed.
Missing BillAmount values are filled.
The dataset contains correct and complete billing information.
Example

Before Cleaning

PatientID	Department	BillAmount	ReceptionistID
101	Cardiology	5000	R01
102	Neurology	NaN	R02
101	Cardiology	5000	R01

After Cleaning

PatientID	Department	BillAmount
101	Cardiology	5000
102	Neurology	6500

2. All missing values handled, merged dataset across PatientID
Meaning

You should:

Replace all missing (NaN) values.
Combine the Patient Dataset and the Billing Dataset using the common column PatientID.
Example

Patient Dataset

PatientID	Name	Department
101	Rahul	Cardiology
102	Priya	Neurology

Billing Dataset

PatientID	BillAmount
101	5000
102	7000

After using:

merged_df = pd.merge(patient_df, billing_df, on="PatientID")

You get:

PatientID	Name	Department	BillAmount
101	Rahul	Cardiology	5000
102	Priya	Neurology	7000

Now all patient and billing details are together in one DataFrame.

3. Ability to perform further analytics on department-wise revenue or doctor performance
Meaning

After cleaning and merging the data, it becomes easy to analyze it.

For example:

Department-wise Revenue
df.groupby("Department")["BillAmount"].sum()

Output:

Department	Total Revenue
Cardiology	50000
Neurology	42000
Orthopedic	38000

This tells you how much money each department earned.

Doctor Performance
df.groupby("Doctor")["BillAmount"].sum()

Output:

Doctor	Total Bill Amount
Dr. Ravi	35000
Dr. Priya	42000

This helps compare doctors based on the total billing associated with their patients.